homeassistant/components/electric_kiwi/sensor.py

```diff
@@ -0,0 +1,140 @@
+        value_func=lambda hop: datetime.combine(
+            datetime.today(), datetime.strptime(hop.start.start_time, "%I:%M %p").time()
+        ).astimezone(dt_util.DEFAULT_TIME_ZONE),
+    ),
+    ElectricKiwiHOPSensorEntityDescription(
+        key=ATTR_EK_HOP_END,
+        name="Hour of free power end",
+        device_class=SensorDeviceClass.TIMESTAMP,
+        value_func=lambda hop: datetime.combine(
+            datetime.today(),
+            datetime.strptime(hop.end.end_time, "%I:%M %p").time(),
+        ).astimezone(dt_util.DEFAULT_TIME_ZONE),
+    ),
+)
+
+
+class ElectricKiwiHOPEntity(CoordinatorEntity, SensorEntity):
+    """Entity object for Electric Kiwi sensor."""
+
+    entity_description: ElectricKiwiHOPSensorEntityDescription
+
+    def __init__(
+        self,
+        hop_coordinator: ElectricKiwiHOPDataCoordinator,
+        description: ElectricKiwiHOPSensorEntityDescription,
+        customer_number: int,
+        connection_id: int,
+    ) -> None:
+        """Entity object for Electric Kiwi sensor."""
+        super().__init__(hop_coordinator)
+        self._hop_coordinator: ElectricKiwiHOPDataCoordinator = hop_coordinator
+        self.entity_description = description
+
+        self.customer_number = customer_number
+        self.connection_id = connection_id
+        self.entity_description = description
+        self._attributes = {
+            ATTR_ATTRIBUTION: ATTRIBUTION,
+        }
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return "_".join(
+            [
+                str(self.customer_number),
+                str(self.connection_id),
+                self.entity_description.key,
+            ]
+        )
+
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        return f"{self.entity_description.name}"
+
+    @property
+    def native_value(self) -> datetime | str | None:
+        """Return the state of the sensor."""
+        value: datetime = self.entity_description.value_func(
+            self._hop_coordinator.get_selected_hop()
+        )
+
+        end_time = datetime.combine(
+            datetime.today(),
+            datetime.strptime(
+                self._hop_coordinator.get_selected_hop().end.end_time, "%I:%M %p"
+            ).time(),
+        ).astimezone(dt_util.DEFAULT_TIME_ZONE)
+
+        if end_time < datetime.now().astimezone(dt_util.DEFAULT_TIME_ZONE):
+            return value + timedelta(days=1)
+        return value
+
+    @property
+    def extra_state_attributes(self) -> dict[str, str]:
+        """Return the state attributes."""
+        return self._attributes
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Update attributes when the coordinator updates."""
+        self._attr_native_value = self.entity_description.value_func(
+            self._hop_coordinator.get_selected_hop()
+        )
+        super()._handle_coordinator_update()
```

Comment:
It is already implementing the `native_property` method, why is the update handler implemented?

Reply:
I'll commit a change that removes this and also tidies up the native_value function as well. I'm not sure on where to place the code that currently resides inside native_value so its a lone function, advice please. Will be pushing all these changes shortly